.travis.yml

```diff
@@ -13,74 +17,111 @@ python:
 
 before_script:
-  # Vagrant is not used on travis: This user is just for the database.
-  - createuser --createdb --no-superuser --no-createrole vagrant
-  - createdb -O vagrant refinery
-
-  - sudo mkdir /data # In vagrant, this is created by puppet; In production, this is an EBS mount.
-  - sudo chown $USER /data
+  - |
+    if ! test -n "$GRUNT_TESTS"; then
+      # Vagrant is not used on travis: This user is just for the database.
+      createuser --createdb --no-superuser --no-createrole vagrant
+      createdb -O vagrant refinery
+      sudo mkdir /data
+      sudo chown $USER /data
+      cd refinery
+      # See http://www.stuartellis.eu/articles/erb/#running-erb-from-the-command-line
+      erb config/config.json.erb > config/config.json
+      python manage.py migrate --noinput
+      cd ../
+    fi
+  - |
+    if ! test -n "$DJANGO_UNIT_TESTS"; then
+      cd refinery
+      npm install -g grunt-cli@0.1.13 bower@1.7.7 --progress false --quiet || ( cat npm-debug.log && false )
+      cd ui
+      npm install --progress false --quiet || ( cat npm-debug.log && false )
+      bower install --config.interactive=false --quiet
+      cd ../../
+    fi
 
-  - cd refinery
-# See http://www.stuartellis.eu/articles/erb/#running-erb-from-the-command-line
-  - erb config/config.json.erb > config/config.json
-  - python manage.py migrate --noinput
-  - npm install -g grunt-cli@0.1.13 bower@1.7.7 --progress false --quiet || ( cat npm-debug.log && false )
-  - cd ui
-  - npm install --progress false --quiet || ( cat npm-debug.log && false )
-  - bower install --config.interactive=false --quiet
-  - cd ../
 script:
-  - set -e # Any error will cause travis to exit early and report a failure.
-
-  # TODO: Not the same behavior as pre-commit hook.
-  # https://github.com/refinery-platform/refinery-platform/issues/1851
-  - flake8 --exclude=migrations,ui ..
+  - "set -e" # Any error will cause travis to exit early and report a failure.
+  - |
+    if ! test -n "$GRUNT_TESTS"; then
+      # TODO: Not the same behavior as pre-commit hook.
+      # https://github.com/refinery-platform/refinery-platform/issues/1851
+      flake8 --exclude=migrations,ui ..
+    fi
 
+  - cd refinery
   - echo 'travis_fold:start:grunt'
-  - pushd ui && grunt && popd
+  - pushd ui
+  - if ! test -n "$DJANGO_UNIT_TESTS"; then grunt make; fi
+
+  - if test -n "$GRUNT_TESTS"; then grunt test; fi
+  - popd
   - echo 'travis_fold:end:grunt'
 
-  - echo 'travis_fold:start:django-tests'
-  - coverage run manage.py test
-  - echo 'travis_fold:end:django-tests'
+  - echo 'travis_fold:start:django-unit-tests'
+  - if test -n "$DJANGO_UNIT_TESTS"; then coverage run manage.py test; fi
+  - echo 'travis_fold:end:django-unit-tests'
+
+  - echo 'travis_fold:start:django-selenium-tests'
+  - if test -n "$DJANGO_SELENIUM_TESTS"; then coverage run manage.py test --pattern="selenium_tests.py"; fi
+  - echo 'travis_fold:end:django-selenium-tests'
```

Comment:
With the build matrix, only one of these sections is being run, so the `travis_fold` stuff should be inside the `if-then`, if it is retained at all.